core/src/main/java/org/apache/iceberg/avro/AvroSchemaWithTypeVisitor.java

```diff
@@ -79,11 +79,18 @@
   private static <T> T visitUnion(Type type, Schema union, AvroSchemaWithTypeVisitor<T> visitor) {
     List<Schema> types = union.getTypes();
     List<T> options = Lists.newArrayListWithExpectedSize(types.size());
+
+    int index = 0;
     for (Schema branch : types) {
       if (branch.getType() == Schema.Type.NULL) {
         options.add(visit((Type) null, branch, visitor));
       } else {
-        options.add(visit(type, branch, visitor));
+        if (AvroSchemaUtil.isOptionSchema(union)) {
+          options.add(visit(type, branch, visitor));
+        } else {
+          options.add(visit(type.asStructType().fields().get(index).type(), branch, visitor));
+        }
+        index++;
```

Comment:
nit: we now maintain index outside the loop, and check if union is option (line 88) every time in the loop.  May be  use one loop for both branches and fields, and re-org to hoist up the union check (if that is feasible)?